docs/pbc/all/order-management-system/202311.0/base-shop/datapayload-conversion/state-machine/common-pitfalls-in-oms-design.md

```diff
@@ -154,6 +168,83 @@ The unused state could have a missing transition.
+
+## Stuck on-enter
+
+**The issue:** In most cases, if you have an order item stuck during the onEnter transition it means you have an unexpected error during execution. I highly recommend checking those issues individually and properly fixing them.
+
+**Solution:** (if you have many of them due to infra, timeouts etc. and you need to fix them ASAP). Create a console command that will trigger onEnter events. I recommend creating a list of OMS states from which you want to check and move order items. Also, set limits for orders (not items!) and time windows. For example, check stuck order items only if they have the last update of the state after 2 hours.
+
+## Saving states (per item transaction)
+
+**Example:** The system has a callback that should move order items from picking started to ready for recalculation. After that, check-condition will move the order further to the recalculation step.
+![saving-states-1](https://github.com/xartsiomx/spryker-docs/assets/110463030/dd060438-7cd2-4453-9cf4-e08cda243003)
+
+**The issue:** During the last transition in the callback from picking finished to ready for recalculation - Jenkins job started the check-condition command. Because of that check-condition takes not all order items from the order and pushes them forward. Only the next job started executing the remaining order items with the delay, because of that many commands (per Order) were triggered twice.
+![saving-states-2](https://github.com/xartsiomx/spryker-docs/assets/110463030/1fd1b30f-00dc-49eb-8d35-37583e140f5e)
+
+**Solution:** This is possible because during the execution of \Spryker\Zed\Oms\Business\OrderStateMachine\OrderStateMachine::saveOrderItems system stores data per item. It was done like this because Core logic expects that you may have more than 1 order in transition. And to not block all of them due to potential failed orders - executing transactions per item. To change that you should first group your order items per order and after that, you can change transaction behavior - to store all order items per only 1 transaction. In this case, check-condition or any other command can’t take order items partially.
+
+## LockedStateMachine
+
+**Example:** In case when multiple processes can push forward an order (callback from third-party & console command) it’s recommended to use LockedStateMachine. Important to understand:
+
+1. It implements the same interface as common StateMachine and has locks for all methods except the check-condition command. The reason behind - you shouldn’t have any conflict inside the execution of the check-condition command.
+
+2. Lock works based on MySQL table spy_state_machine_lock - and because it is MySQL sometimes you can face deadlocks (you have to handle them properly) and it takes more time to work with than memory storage (redis for example). Also locking works on the order item level, but in most cases more efficient will be used on the order level.
```

Comment:
Rewrite this point by splitting into several sentences. Write clear and simple statements. Things unclear:
* How to handle deadlocks properly?
* What takes more time to work with? 
* What do you mean by `work with`?